packages/core/src/Repo.ts

```diff
@@ -0,0 +1,439 @@
+﻿import { RepoHistory } from './types'
+import A from 'automerge'
+import { newid } from 'cevitxe-signal-client'
+import debug from 'debug'
+import { EventEmitter } from 'events'
+import * as idb from 'idb/with-async-ittr-cjs'
+import { ChangeSet, RepoSnapshot } from './types'
+import { DELETED } from './collection'
+import Cache from 'lru-cache'
+
+const DB_VERSION = 1
+export type RepoEventHandler<T> = (documentId: string, doc: A.Doc<T>) => void | Promise<void>
+
+/**
+ *
+ * ### Storage schema
+ *
+ * We use a single database with two object stores: `feeds`, containing changesets in sequential
+ * order, indexed by documentId; and `snapshots`, containing an actual
+ *
+ * There is one repo (and one database) per discovery key.
+ * ```
+ * cevitxe::grid::fancy-lizard (DB)
+ *   feeds (object store)
+ *     1: { id:1, documentId: abc123, changeSet: [...]}
+ *     2: { id:2, documentId: abc123, changeSet: [...]}
+ *     3: { id:3, documentId: abc123, changeSet: [...]}
+ *     4: { id:4, documentId: qrs567, changeSet: [...]}
+ *     5: { id:5, documentId: qrs567, changeSet: [...]}
+ *     6: { id:6, documentId: qrs567, changeSet: [...]}
+ *   snapshots (object store)
+ *     abc123: [snapshot]
+ *     qrs567: [snapshot]
+ * ```
+ */
+export class Repo<T = any> extends EventEmitter {
+  private log: debug.Debugger
+
+  /**
+   * The discovery key is a unique ID for this dataset, used to identify it when seeking peers with
+   * whom to synchronize. In the example apps we use randomly generated two-word names like
+   * `golden-lizard`. It could also be a UUID.
+   */
+  private discoveryKey: string
+
+  /**
+   * Name to distinguish this application's data from others that this browser might have stored;
+   * e.g. `grid` or `todos`.
+   */
+  public databaseName: string
+
+  public clientId: string
+  /**
+   * In-memory map of document snapshots.
+   */
+  private state: RepoSnapshot<T>
+
+  private docCache: Cache<string, any>
+
+  /**
+   * Document change event listeners. Each handler fires every time a document is set or removed.
+   */
+  private handlers: Set<RepoEventHandler<T>>
+
+  private database?: idb.IDBPDatabase<unknown>
+
+  constructor(discoveryKey: string, databaseName: string, clientId: string = newid()) {
+    super()
+    this.discoveryKey = discoveryKey
+    this.databaseName = databaseName
+    this.log = debug(`cevitxe:repo:${databaseName}`)
+    this.state = {}
+    this.handlers = new Set()
+    this.docCache = new Cache({ max: 1000 })
+    this.clientId = clientId
+  }
+
+  // PUBLIC METHODS
+
+  /**
+   * Opens the local database and returns a reference to it.
+   * @returns An `idb` wrapper for an indexed DB.
+   */
+  async open() {
```

Comment:
Since this is 100% required (it's the async part of construction), I suggest we add a flag (set at the end of `open`) and a helper method that checks it and throws if the repo isn't open yet. All the instance methods would call it to avoid hard-to-track bugs.

Reply:
👍 I've added a TODO
69592d60a4d05244565d09d70541da9299d376bd